### ISSUE TITLE
feat: Compatibility Matrix Generator - model/optimizer/loss function compatibility

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#963]
## Parent Epic
Part of #956 — Model Domain Attributes

## Overview
Create a Roslyn incremental source generator that analyzes generic constraints, interface implementations, and attribute metadata to produce a compile-time compatibility matrix showing which models work with which optimizers, loss functions, and preprocessors.

## Problem Statement
Currently, users discover incompatibilities at runtime when they pair a model with an incompatible optimizer or loss function. This generator shifts that discovery to compile time by analyzing type constraints and producing a queryable compatibility matrix.

### Input Sources
1. **Generic constraints** — e.g., `where T : IFloatingPoint<T>` limits which numeric types work
2. **Interface implementations** — e.g., `ISupportsRegularization<T>` indicates optimizer compatibility
3. **`[ModelCategory]` attribute** — classifiers need classification loss functions, regressors need regression losses
4. **`[ModelTask]` attribute** — certain tasks imply certain preprocessors
5. **Constructor parameters** — required dependencies reveal what components a model needs

### Output: `ModelCompatibility.g.cs`

```csharp
// Auto-generated
public static partial class ModelCompatibility
{
    /// <summary>
    /// Returns optimizers compatible with the given model type.
    /// </summary>
    public static IReadOnlyList<Type> GetCompatibleOptimizers(Type modelType);
    
    /// <summary>
    /// Returns loss functions compatible with the given model type.
    /// </summary>
    public static IReadOnlyList<Type> GetCompatibleLossFunctions(Type modelType);
    
    /// <summary>
    /// Returns preprocessors recommended for the given model type.
    /// </summary>
    public static IReadOnlyList<Type> GetRecommendedPreprocessors(Type modelType);
    
    /// <summary>
    /// Checks if a model and optimizer are compatible.
    /// </summary>
    public static bool IsCompatible(Type modelType, Type optimizerType);
    
    /// <summary>
    /// Returns all compatibility information for a model.
    /// </summary>
    public static ModelCompatibilityInfo GetInfo(Type modelType);
}

public class ModelCompatibilityInfo
{
    public Type ModelType { get; }
    public IReadOnlyList<Type> CompatibleOptimizers { get; }
    public IReadOnlyList<Type> CompatibleLossFunctions { get; }
    public IReadOnlyList<Type> RecommendedPreprocessors { get; }
    public IReadOnlyList<string> Warnings { get; }
}
```

### Compatibility Rules

| Model Category | Compatible Loss Functions | Compatible Optimizers |
|---|---|---|
| Classifier | CrossEntropy, BinaryCrossEntropy, Hinge, FocalLoss | All gradient-based |
| Regression | MSE, MAE, Huber, LogCosh | All gradient-based |
| GAN | Adversarial, Wasserstein, HingeLoss | Adam, RMSProp (not SGD recommended) |
| Autoencoder | MSE, BCE, KLDivergence | All gradient-based |
| TimeSeries | MSE, MAE, QuantileLoss | Adam, AdaGrad |
| SurvivalModel | CoxPartialLikelihood, LogRankLoss | Adam, L-BFGS |

### Diagnostic Warnings

The generator should also emit warnings (not errors) for suspicious combinations:
- `AIDNT001`: Model uses `[ModelCategory(Classifier)]` but paired with regression loss function
- `AIDNT002`: Model uses `[ModelCategory(GAN)]` but optimizer is plain SGD
- `AIDNT003`: Model requires specific preprocessor but none detected in pipeline

## Implementation Notes

- Use `IIncrementalGenerator` for performance
- Analyze `SyntaxTree` and `SemanticModel` for generic constraints and interfaces
- Cross-reference with attribute data from `[ModelCategory]` and `[ModelTask]`
- Build lookup dictionaries at compile time, emit as static readonly collections
- net471 compatible output

## Acceptance Criteria
- [ ] Generator produces `ModelCompatibility.g.cs`
- [ ] `GetCompatibleOptimizers()` returns correct types based on model category
- [ ] `GetCompatibleLossFunctions()` returns correct types based on model category
- [ ] `IsCompatible()` correctly validates model+optimizer pairs
- [ ] Diagnostic warnings emitted for suspicious combinations
- [ ] Builds on both net10.0 and net471

## Dependencies
- #957 (attributes and enums must exist first)
- #958 (models must be annotated)
- #960 (registry generator provides model metadata at compile time)
